log/log.go

```diff
@@ -121,27 +210,57 @@ func (hook ContextHook) Levels() []logrus.Level {
 	return logrus.AllLevels
 }
 
+func fmtCaller(frame runtime.Frame) string {
+	return fmt.Sprintf(
+		"%s:%d>%s",
+		path.Base(frame.File),
+		frame.Line,
+		path.Base(frame.Function),
+	)
+}
+
 func (hook ContextHook) Fire(entry *logrus.Entry) error {
-	//'skip' = 6 is the default call stack skip, which
-	//works ootb when Error(), Warn(), etc. are called
-	//for Errorf(), Warnf(), etc. - we have to skip 1 lvl up
-	for skip := 6; skip < 8; skip++ {
-		if pc, file, line, ok := runtime.Caller(skip); ok {
-			funcName := runtime.FuncForPC(pc).Name()
-
-			//detect if we're still in logrus (formatting funcs)
-			if !strings.Contains(funcName, "github.com/sirupsen/logrus") {
-				entry.Data["file"] = path.Base(file)
-				entry.Data["func"] = path.Base(funcName)
-				entry.Data["line"] = line
+	const (
+		minCallDepth = 6 // logrus.Logger.Log
+		maxCallDepth = 8 // logrus.Logger.<Level>f
+	)
+	var pcs [1 + maxCallDepth - minCallDepth]uintptr
+	if _, ok := entry.Data["caller"]; !ok {
+		// We don't know how deep we are in the callstack since the hook can be fired
+		// at different levels. Search between depth 6 -> 8.
+		i := runtime.Callers(minCallDepth, pcs[:])
+		frames := runtime.CallersFrames(pcs[:i])
+		var caller *runtime.Frame
+		for frame, ok := frames.Next(); ok; frame, ok = frames.Next() {
+			if !strings.HasPrefix(frame.Function, "github.com/sirupsen/logrus") {
+				caller = &frame
```

Comment:
I know I am annoying, but could we please define `"caller"` and `"github.com/sirupsen/logrus"`?